### PR TITLE
New programming model

### PR DESCRIPTION
**Pull request by andystaples**
### Issue describing the changes in this PR

Adds support for worker indexing to the functions worker

### Pull request checklist

* [ ] My changes **do not** require documentation changes
  * [ ] Otherwise: Documentation issue linked to PR
* [ ] My changes **should not** be added to the release notes for the next release
  * [ ] Otherwise: I've added my notes to `release_notes.md`
* [x] My changes **do not** need to be backported to a previous version
  * [ ] Otherwise: Backport tracked by issue/PR #issue_or_pr
* [ ] I have added all required tests (Unit tests, E2E tests)

### CODE REVIEW

**Comment by andystaples** [comment]
@Francisco-Gamino I think this PR is correct. The rebase was awful and doubled the amount of commits, not sure why. Lots still to do here, gotta bundle the module with the worker and figure out how to get the cmdlet available in order to write tests for the new programming model

File: src/Microsoft.Azure.Functions.PowerShellWorker.csproj [hunk 1]
```diff
@@ -35,4 +35,8 @@ Licensed under the MIT license. See LICENSE file in the project root for full li
       <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
     </Content>
   </ItemGroup>
+
+  <ItemGroup>
```
Comment (Francisco-Gamino) [inline src/Microsoft.Azure.Functions.PowerShellWorker.csproj]: Please remove this entry.

File: src/RequestProcessor.cs [hunk 2]
```diff
@@ -153,6 +182,7 @@ internal StreamingMessage ProcessFunctionLoadRequest(StreamingMessage request)
             stopwatch.Start();
 
             FunctionLoadRequest functionLoadRequest = request.FunctionLoadRequest;
+            string lrs = System.Text.Json.JsonSerializer.Serialize<FunctionLoadRequest>(functionLoadRequest);
```
Comment (Francisco-Gamino) [inline src/RequestProcessor.cs]: Could you please add a comment on what this variable is used for?

File: src/RequestProcessor.cs [hunk 3]
```diff
@@ -30,9 +32,9 @@ internal class RequestProcessor
 
         // Holds the exception if an issue is encountered while processing the function app dependencies.
         private Exception _initTerminatingError;
-
-        // Indicate whether the FunctionApp has been initialized.
-        private bool _isFunctionAppInitialized;
+        
+        // Indicate whether the dependencies have been installed
+        private bool _areDependenciesInstalled;
```
Comment (Francisco-Gamino) [inline src/RequestProcessor.cs]: We should keep `_isFunctionAppInitialized` given that an app could have dependencies or these might be bundled with the function app content.

In this case, `_isFunctionAppInitialized` is used to indicate whether the function app is ready for a function execution.

File: src/RequestProcessor.cs [hunk 4]
```diff
@@ -117,6 +124,28 @@ internal StreamingMessage ProcessWorkerInitRequest(StreamingMessage request)
                 RemoteSessionNamedPipeServer.CreateCustomNamedPipeServer(pipeName);
             }
 
+            try
+            {
```
Comment (Francisco-Gamino) [inline src/RequestProcessor.cs:133]: I believe that you moved the installation of function app dependencies to `ProcessWorkerInitRequest`. However, the Functions Host has a `60 seconds` timeout for the worker response, so this will cause a timeout exception. Could you please ping the Team channel on why we need this change? 

Alternately, if the function app does not have any dependencies installed, we can start an async task in `ProcessWorkerInitRequest` to install the dependencies and send the response back to the Function Host.  Then, on the first function invocation, we can check if the task completed executing, if this is the case, we can continue with function invocation. Otherwise, wait until the task completes. What are your thoughts on this @michaelpeng36 @AnatoliB @davidmrdavid ?

File: src/RequestProcessor.cs [hunk 5]
```diff
@@ -240,6 +264,7 @@ internal StreamingMessage ProcessInvocationRequest(StreamingMessage request)
         {
             try
             {
+                string ins = System.Text.Json.JsonSerializer.Serialize<InvocationRequest>(request.InvocationRequest);
```
Comment (Francisco-Gamino) [inline src/RequestProcessor.cs]: Could you please add a comment on what this variable is used for?

File: src/WorkerIndexing/BindingInformation.cs [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+        public Directions Direction { get; set; } = Directions.Unknown;
+        public string Type { get; set; } = "";
+        public string Name { get; set; } = "";
+        public Dictionary<string, Object> otherInformation { get; set; } = new Dictionary<string, Object>();
+
+        internal string ConvertToRpcRawBinding(out BindingInfo bindingInfo)
+        {
+            string rawBinding = string.Empty;
+            JObject rawBindingObject = new JObject();
+            rawBindingObject.Add("name", Name);
```
Comment (Francisco-Gamino) [inline src/WorkerIndexing/BindingInformation.cs]: Please make `"name"` a constant.

File: src/WorkerIndexing/FunctionInformation.cs [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+
+        internal RpcFunctionMetadata ConvertToRpc()
+        {
+            RpcFunctionMetadata returnMetadata = new RpcFunctionMetadata();
+            returnMetadata.FunctionId = FunctionId;
+            returnMetadata.Directory = Directory;
+            returnMetadata.EntryPoint = EntryPoint;
+            returnMetadata.Name = Name;
+            returnMetadata.ScriptFile = ScriptFile;
+            returnMetadata.Language = "powershell";
```
Comment (Francisco-Gamino) [inline src/WorkerIndexing/FunctionInformation.cs]: Please make `"powershell"` a constant.

File: src/WorkerIndexing/WorkerIndexingHelper.cs [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+        {
+            List<RpcFunctionMetadata> indexedFunctions = new List<RpcFunctionMetadata>();
+
+            InitialSessionState initial = InitialSessionState.CreateDefault();
+            Runspace runspace = RunspaceFactory.CreateRunspace(initial);
+            runspace.Open();
+            System.Management.Automation.PowerShell ps = System.Management.Automation.PowerShell.Create();
+            ps.Runspace = runspace;
+
+            ps.AddCommand("Get-FunctionsMetadata").AddArgument(baseDir);
```
Comment (Francisco-Gamino) [inline src/WorkerIndexing/WorkerIndexingHelper.cs]: Please make this a constant. Here is an example: https://github.com/Azure/azure-functions-powershell-worker/blob/ac95680f3d905a85352410d942f66e536d9ddde5/src/DependencyManagement/PowerShellGalleryModuleProvider.cs#L82

File: src/worker.config.json [hunk 9]
```diff
@@ -1,11 +1,12 @@
 {
-    "description":{
-        "language":"powershell",
-        "extensions":[".ps1", ".psm1"],
-        "defaultExecutablePath":"dotnet",
-        "defaultWorkerPath":"%FUNCTIONS_WORKER_RUNTIME_VERSION%/Microsoft.Azure.Functions.PowerShellWorker.dll",
-        "supportedRuntimeVersions":["7", "7.2"],
+    "description": {
```
Comment (Francisco-Gamino) [inline src/worker.config.json:2]: We should not make any changes to the existing `worker.config.json`, so we can continue to have the existing behavior. However, could you please document the required changes in `worker.config.json` to enable `workerIndexing`? Perhaps the `src/WorkerIndexing/WorkerIndexingHelper.cs` would be a good place for this. Let me know what you think.

**Review by Francisco-Gamino — COMMENTED** [review]
Thank you @andystaples for sending this out! I've left a few comments. Let me know if you have any questions.

File: src/Microsoft.Azure.Functions.PowerShellWorker.csproj (hunk 1 above)
Comment (andystaples) [reply src/Microsoft.Azure.Functions.PowerShellWorker.csproj]: Done

File: src/RequestProcessor.cs (hunk 2 above)
Comment (andystaples) [reply src/RequestProcessor.cs]: I was debugging with this, forgot to take it out. Removed

File: src/RequestProcessor.cs (hunk 3 above)
Comment (andystaples) [reply src/RequestProcessor.cs]: OK, changing it back

File: src/RequestProcessor.cs (hunk 4 above)
Comment (andystaples) [reply src/RequestProcessor.cs:133]: To clarify, the dependency download actually hasn't moved. Before and after my changes, the download happens on the first function load request. I have just moved some of the pre-initialization to the function load request, so that certain properties (like the functions app directory) are available during the metadata request which happens before function load requests. 

The worker already does kick off a task to install dependencies on the first function load request, it does not wait until all dependencies are installed to return.

File: src/RequestProcessor.cs (hunk 5 above)
Comment (andystaples) [reply src/RequestProcessor.cs]: Another debug string that I forgot to remove

File: src/WorkerIndexing/BindingInformation.cs (hunk 6 above)
Comment (andystaples) [reply src/WorkerIndexing/BindingInformation.cs]: Done

File: src/WorkerIndexing/FunctionInformation.cs (hunk 7 above)
Comment (andystaples) [reply src/WorkerIndexing/FunctionInformation.cs]: Done

File: src/WorkerIndexing/WorkerIndexingHelper.cs (hunk 8 above)
Comment (andystaples) [reply src/WorkerIndexing/WorkerIndexingHelper.cs]: Done

File: src/worker.config.json (hunk 9 above)
Comment (andystaples) [reply src/worker.config.json:2]: As per our conversation earlier this morning, two flags are necessary in order for the host to attempt worker indexing - the first is in the worker.config.json, and the second is in the local.settings.json/function app environment variables. Having this string in worker.config.json reports to the host that the worker is compatible with worker indexing, and the one in local.settings.json reports that the app is compatible. 
So, long story short, making this change doesn't harm existing apps.

File: src/RequestProcessor.cs (hunk 4 above)
Comment (Francisco-Gamino) [reply src/RequestProcessor.cs:133]: Thank you @andystaples for the clarification. 

Yes, in the first function invocation, we use the first PowerShell instance to download the function app dependencies and wait until the installation completes before invoking the function. The wait takes place here https://github.com/Azure/azure-functions-powershell-worker/blob/187ad3369a930f9f0acda258fa58618bf9059a27/src/DependencyManagement/DependencyManager.cs#L164

File: src/WorkerIndexing/WorkerIndexingHelper.cs [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+            ps.AddCommand(GetFunctionsMetadataCmdletName).AddArgument(baseDir);
+            string outputString = string.Empty;
+            foreach (PSObject rawMetadata in ps.Invoke())
+            {
+                if (outputString != string.Empty)
+                {
+                    throw new Exception(PowerShellWorkerStrings.GetFunctionsMetadataMultipleResultsError);
+                }
+                outputString = rawMetadata.ToString();
+                //Console.WriteLine(rawMetadata.ToString());
```
Comment (Francisco-Gamino) [inline src/WorkerIndexing/WorkerIndexingHelper.cs]: Please remove this if it is not being used.

File: src/WorkerIndexing/WorkerIndexingHelper.cs [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+using System;
+using System.Collections.Generic;
+using System.Management.Automation;
+using System.Management.Automation.Runspaces;
+
+namespace Microsoft.Azure.Functions.PowerShellWorker.WorkerIndexing
+{
+    internal class WorkerIndexingHelper
+    {
+        const string GetFunctionsMetadataCmdletName = "Get-FunctionsMetadata";
```
Comment (Francisco-Gamino) [inline src/WorkerIndexing/WorkerIndexingHelper.cs:19]: Please include the module name, e.g., `"AzureFunctions.PowerShell.SDK\\Get-FunctionsMetadata"`

File: src/resources/PowerShellWorkerStrings.resx [hunk 12]
```diff
@@ -352,4 +352,10 @@
   <data name="DependencySnapshotDoesNotContainAcceptableModuleVersions" xml:space="preserve">
     <value>Dependency snapshot '{0}' does not contain acceptable module versions.</value>
   </data>
+  <data name="GetFunctionsMetadataMultipleResultsError" xml:space="preserve">
```
Comment (Francisco-Gamino) [inline src/resources/PowerShellWorkerStrings.resx:355]: Could you please ping the Team channel to get feedback on the new string resources you added?

File: src/WorkerIndexing/BindingInformation.cs [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+            outInfo.Direction = (BindingInfo.Types.Direction)Direction;
+            rawBindingObject.Add(BindingDirectionKey, Enum.GetName(typeof(BindingInfo.Types.Direction), outInfo.Direction).ToLower());
+            outInfo.Type = Type;
+            rawBindingObject.Add(BindingTypeKey, Type);
+
+            foreach (KeyValuePair<string, Object> pair in otherInformation)
+            {
+                // Wow this sucks, lots of overserialization
+                rawBindingObject.Add(pair.Key, JToken.FromObject(pair.Value));
+                //Console.WriteLine(pair.Key);
```
Comment (Francisco-Gamino) [inline src/WorkerIndexing/BindingInformation.cs]: Please remove this if it is not needed.

File: src/WorkerIndexing/BindingInformation.cs [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+                throw new Exception(PowerShellWorkerStrings.InvalidBindingInfoDirection);
+            }
+            outInfo.Direction = (BindingInfo.Types.Direction)Direction;
+            rawBindingObject.Add(BindingDirectionKey, Enum.GetName(typeof(BindingInfo.Types.Direction), outInfo.Direction).ToLower());
+            outInfo.Type = Type;
+            rawBindingObject.Add(BindingTypeKey, Type);
+
+            foreach (KeyValuePair<string, Object> pair in otherInformation)
+            {
+                // Wow this sucks, lots of overserialization
```
Comment (Francisco-Gamino) [inline src/WorkerIndexing/BindingInformation.cs]: If you think that the overserialization issue can be improved, please add some suggestions in the comment.

File: src/RequestProcessor.cs (hunk 4 above)
Comment (andystaples) [reply src/RequestProcessor.cs:133]: That method waits for the dependency install task to complete, but that task is kicked off on the first function load request, using the method StartDependencyInstallationIfNeeded on line 230 of RequestProcessor.cs. Anyway, I think we're good here

File: src/WorkerIndexing/WorkerIndexingHelper.cs (hunk 11 above)
Comment (andystaples) [reply src/WorkerIndexing/WorkerIndexingHelper.cs:19]: Done

File: src/WorkerIndexing/BindingInformation.cs (hunk 13 above)
Comment (andystaples) [reply src/WorkerIndexing/BindingInformation.cs]: Done

File: src/WorkerIndexing/BindingInformation.cs (hunk 14 above)
Comment (andystaples) [reply src/WorkerIndexing/BindingInformation.cs]: I don't have any ideas, I'm removing the comment

File: src/WorkerIndexing/WorkerIndexingHelper.cs (hunk 10 above)
Comment (andystaples) [reply src/WorkerIndexing/WorkerIndexingHelper.cs]: Done

File: src/Microsoft.Azure.Functions.PowerShellWorker.csproj [hunk 15]
```diff
@@ -1,4 +1,4 @@
-<!--
+﻿<!--
```
Comment (davidmrdavid) [inline src/Microsoft.Azure.Functions.PowerShellWorker.csproj]: nipickt: did anything change in this file? Seems like nothing changed. If so - can we instruct git to "revert this file" so that it doesn't register a change in the PR?

File: src/RequestProcessor.cs [hunk 16]
```diff
@@ -117,6 +124,28 @@ internal StreamingMessage ProcessWorkerInitRequest(StreamingMessage request)
...
+                _dependencyManager = new DependencyManager(request.WorkerInitRequest.FunctionAppDirectory, logger: rpcLogger);
+
+                _powershellPool.Initialize(_firstPwshInstance);
+
+                rpcLogger.Log(isUserOnlyLog: false, LogLevel.Trace, string.Format(PowerShellWorkerStrings.FirstFunctionLoadCompleted, stopwatch.ElapsedMilliseconds));
+            }
+            catch (Exception e)
+            {
+                // Failure that happens during this step is terminating and we will need to return a failure response to
+                // all subsequent 'FunctionLoadRequest'. Cache the exception so we can reuse it in future calls.
```
Comment (davidmrdavid) [inline src/RequestProcessor.cs]: A suggestion on wording:
```suggestion
                // This is a terminating failure: we will need to return a failure response to
                // all subsequent 'FunctionLoadRequest'. Cache the exception so we can reuse it in future calls.
```

File: src/RequestProcessor.cs [hunk 17]
```diff
@@ -117,6 +124,28 @@ internal StreamingMessage ProcessWorkerInitRequest(StreamingMessage request)
...
+            {
+                var rpcLogger = new RpcLogger(_msgStream);
+                rpcLogger.SetContext(request.RequestId, null);
+
+                _dependencyManager = new DependencyManager(request.WorkerInitRequest.FunctionAppDirectory, logger: rpcLogger);
+
+                _powershellPool.Initialize(_firstPwshInstance);
+
+                rpcLogger.Log(isUserOnlyLog: false, LogLevel.Trace, string.Format(PowerShellWorkerStrings.FirstFunctionLoadCompleted, stopwatch.ElapsedMilliseconds));
+            }
```
Comment (davidmrdavid) [inline src/RequestProcessor.cs:142]: Could you please add a comment here on what we're trying to achieve, as well as what kind of error we're protecting against in the try-catch?

File: src/RequestProcessor.cs [hunk 18]
```diff
@@ -340,6 +363,19 @@ internal StreamingMessage ProcessInvocationCancelRequest(StreamingMessage reques
...
+        
+        private StreamingMessage ProcessFunctionMetadataRequest(StreamingMessage request)
+        {
+            // WorkerStatusResponse type says that it is not used but this will create an empty one anyway to return to the host
+            StreamingMessage response = NewStreamingMessageTemplate(
+                request.RequestId,
+                StreamingMessage.ContentOneofCase.FunctionMetadataResponse,
+                out StatusResult status);
+
+            response.FunctionMetadataResponse.FunctionMetadataResults.AddRange(WorkerIndexingHelper.IndexFunctions(request.FunctionsMetadataRequest.FunctionAppDirectory));
```
Comment (davidmrdavid) [inline src/RequestProcessor.cs]: Without knowing the specifics of the contract between the Host and worker, I can't decipher what the comment at the top of this block is trying to say. Can you please add a bit more context? What exactly "says that it is not used"?, and what is "it" in this context? Thanks!

File: src/WorkerIndexing/WorkerIndexingHelper.cs [hunk 19]
```diff
@@ -0,0 +1,50 @@
...
+using System;
+using System.Collections.Generic;
+using System.Management.Automation;
+using System.Management.Automation.Runspaces;
+
+namespace Microsoft.Azure.Functions.PowerShellWorker.WorkerIndexing
+{
+    internal class WorkerIndexingHelper
+    {
+        const string GetFunctionsMetadataCmdletName = "AzureFunctions.PowerShell.SDK\\Get-FunctionsMetadata";
```
Comment (davidmrdavid) [inline src/WorkerIndexing/WorkerIndexingHelper.cs]: Should we move this string to the strings-resource file?

File: src/WorkerIndexing/WorkerIndexingHelper.cs [hunk 20]
```diff
@@ -0,0 +1,50 @@
...
+            List<RpcFunctionMetadata> indexedFunctions = new List<RpcFunctionMetadata>();
+
+            InitialSessionState initial = InitialSessionState.CreateDefault();
+            Runspace runspace = RunspaceFactory.CreateRunspace(initial);
+            runspace.Open();
+            System.Management.Automation.PowerShell ps = System.Management.Automation.PowerShell.Create();
+            ps.Runspace = runspace;
+
+            ps.AddCommand(GetFunctionsMetadataCmdletName).AddArgument(baseDir);
+            string outputString = string.Empty;
```
Comment (davidmrdavid) [inline src/WorkerIndexing/WorkerIndexingHelper.cs]: I think this block would benefit from a small comment explaining what data your runspace is achieving, and how exactly this is helping the indexing process.

**Review by davidmrdavid — COMMENTED** [review]
Some minor thoughts from skimming the PR. Looking good!

File: src/RequestProcessor.cs (hunk 16 above)
Comment (andystaples) [reply src/RequestProcessor.cs]: Added

File: src/RequestProcessor.cs (hunk 17 above)
Comment (andystaples) [reply src/RequestProcessor.cs:142]: Done

File: src/RequestProcessor.cs (hunk 18 above)
Comment (andystaples) [reply src/RequestProcessor.cs]: I have no idea what this comment is about either. My guess - some misguided note I made while working on my first prototype and forgot to remove. I have removed it

File: src/WorkerIndexing/WorkerIndexingHelper.cs (hunk 19 above)
Comment (andystaples) [reply src/WorkerIndexing/WorkerIndexingHelper.cs]: This is the way that other places in the code which call cmdlets deal with it. Strings-resources are for customer-facing error/warning messages. I propose creating a separate issue on the worker to track all occurrences and move them to some global constants file.

File: src/WorkerIndexing/WorkerIndexingHelper.cs (hunk 20 above)
Comment (andystaples) [reply src/WorkerIndexing/WorkerIndexingHelper.cs]: Done. This is a much larger issue, actually, and without a better approach, will lead to the AzureFunctions.PowerShell.SDK module never being available to the worker in Azure, even if we bundle it with the worker. I have left a long comment explaining the issue and some proposed solutions.

File: src/resources/PowerShellWorkerStrings.resx [hunk 21]
```diff
@@ -352,4 +352,10 @@
...
+    <value>Multiple results from metadata cmdlet</value>
+  </data>
+  <data name="InvalidBindingInfoDirection" xml:space="preserve">
+    <value>The bindingInfo's Direction is not valid</value>
```
Comment (michaelpeng36) [inline src/resources/PowerShellWorkerStrings.resx]: Would we be able to specify which binding name this corresponds to?

Comment (andystaples) [reply src/resources/PowerShellWorkerStrings.resx]: Yes

**Review by michaelpeng36 — APPROVED** [review]

File: src/WorkerIndexing/BindingInformation.cs [hunk 22]
```diff
@@ -0,0 +1,54 @@
+﻿using Microsoft.Azure.WebJobs.Script.Grpc.Messages;
```
Comment (AnatoliB) [inline src/WorkerIndexing/BindingInformation.cs]: Please add the standard copyright/license header to all the new files - just copy it from other files.

Comment (andystaples) [reply src/WorkerIndexing/BindingInformation.cs]: Done
